flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java

```diff
@@ -683,21 +665,66 @@ public void testJobAlreadyDone() throws Exception {
                     jobManagerRunner.getResultFuture();
 
             JobManagerRunnerResult result = resultFuture.get();
-            assertEquals(
-                    JobStatus.FAILED,
-                    result.getExecutionGraphInfo().getArchivedExecutionGraph().getState());
+            assertThat(result.getExecutionGraphInfo().getArchivedExecutionGraph().getState())
+                    .isEqualTo(JobStatus.FAILED);
         }
     }
 
+    @Test
+    void testJobMasterServiceLeadershipRunnerCloseWhenElectionServiceGrantLeaderShip()
+            throws Exception {
+        final TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory
+                testingLeaderElectionDriverFactory =
+                        new TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory();
+        final LeaderElectionService defaultLeaderElectionService =
+                new DefaultLeaderElectionService(testingLeaderElectionDriverFactory);
+
+        final JobMasterServiceLeadershipRunner jobManagerRunner =
+                newJobMasterServiceLeadershipRunnerBuilder()
+                        .setJobMasterServiceProcessFactory(
+                                TestingJobMasterServiceProcessFactory.newBuilder().build())
+                        .setLeaderElectionService(defaultLeaderElectionService)
+                        .build();
+
+        jobManagerRunner.start();
+        final TestingLeaderElectionDriver currentLeaderDriver =
+                Preconditions.checkNotNull(
+                        testingLeaderElectionDriverFactory.getCurrentLeaderDriver());
+
+        final CheckedThread contenderCloseThread =
+                new CheckedThread() {
+                    @Override
+                    public void go() {
+                        try {
+                            jobManagerRunner.close();
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                };
+        contenderCloseThread.start();
+
+        // grant leadership.
+        currentLeaderDriver.isLeader();
```

Comment:
This test is not necessarily failing when running it without the fix you're proposing in this PR. The `close()` call of `jobManagerRunner` needs to be halted to influence the concurrent processing of the `currentLeaderDriver.isLeader()` and the `jobManagerRunner.close()` method. Ideally, this method should fail/block consistently without the fix but should succeed with the fix.
I guess, you could achieve that by inducing your own `TestingJobMasterServiceProcess` through the `TestingJobMasterServiceProcessFactory` in [line 685](https://github.com/apache/flink/pull/21137/files#diff-c010bcd788571261ea6c43a17ad7aaea719a64a7b8a080d76939e7dbbc5a4b32R685). This `TestingJobMasterServiceProcess` gives you control over the `CompletableFuture` that's returned by [JobMasterServiceProcess.closeAsync() in JobMasterServiceLeadershipRunner.closeAsync()](https://github.com/apache/flink/blob/bfe4f9cc3d67d37a2258ab4226d70b6a7d24f22c/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L145). The `JobMasterServiceLeadershipRunner` will only proceeed with stopping the `DefaultLeaderElectionService` if this `CompletableFuture` completes. WDYT?

Reply:
@XComp Thank you for your advice. 
At the beginning, I didn't think of a good way to control the completion of completableFuture, and the probability of deadlock reproduce in my local environment is very high. But your suggestion did find a good control injection point, so i decided to test it in this way. 
But if we want to reproduce the problem 100%, it seems that we need another control point. Maybe `TestingLeaderElectionDriver#isLeader` is a suitable choice. What's your opinion?

Reply:
@XComp Thank you for your advice.
At the beginning, I didn't think of a good way to control the completion of completableFuture, and the probability of deadlock reproduce in my local environment is very high. But your suggestion did find a good control injection point, so i decided to test it in this way.
But if we want to reproduce the problem 100%, it seems that we need another control point. Maybe TestingLeaderElectionDriver#isLeader is a suitable choice. What's your opinion?

Reply:
I went ahead and created a PR with a running example. See reswqa/flink#1 and feel free to edit it. Please see the PRs description. It's not meant to be merged as is but rather an example on how you would be able to test it. I'm happy to discuss any questions you have. But I hope the inline comments are helpful already.